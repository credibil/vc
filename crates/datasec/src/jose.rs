//! # JSON Object Signing and Encryption (JOSE) Proofs
//!
//! [JOSE] proofs are enveloping proofs for Credentials based on JWT [RFC7519],
//! JWS [RFC7515], and JWK [RFC7517].
//!
//! The Securing Verifiable Credentials using JOSE and COSE [VC-JOSE-COSE]
//! recommendation defines a "bridge" between these and the Verifiable
//! Credentials Data Model v2.0, specifying the suitable header claims, media
//! types, etc.
//!
//! In the case of JOSE, the Credential is the "payload". This is preceded by a
//! suitable header whose details are specified by Securing Verifiable
//! Credentials using JOSE and COSE for the usage of JWT. These are encoded,
//! concatenated, and signed, to be transferred in a compact form by one entity
//! to an other (e.g., sent by the holder to the verifier). All the intricate
//! details on signatures, encryption keys, etc., are defined by the IETF
//! specifications; see Example 6 for a specific case.
//!
//! ## Note
//!
//! If the JWT is only a JWE, iss, exp and aud MUST be omitted in the JWT Claims
//! Set of the JWE, and the processing rules as per JARM Section 2.4 related to
//! these claims do not apply. [OpenID4VP] JWT - JWE
//!
//! ```json
//! {
//!   "vp_token": "eyJI...",
//!   "presentation_submission": {...}
//! }
//! ```
//!
//! [JOSE]: https://datatracker.ietf.org/wg/jose/about
//! [RFC7515]: https://www.rfc-editor.org/rfc/rfc7515
//! [RFC7517]: https://www.rfc-editor.org/rfc/rfc7517
//! [RFC7519]: https://www.rfc-editor.org/rfc/rfc7519
//! [VC-JOSE-COSE]: https://w3c.github.io/vc-jose-cose
//! [OpenID4VP]: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html

pub mod jwa;
pub mod jwe;
pub mod jwk;
pub mod jws;
pub mod jwt;
