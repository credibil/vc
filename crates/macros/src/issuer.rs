use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::parse::{Error, Result};

use crate::parse::Data;

pub fn create_offer_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let subject_id = fields.remove("subject_id").unwrap();
    let credential_configuration_ids = fields.remove("credential_configuration_ids").unwrap();
    let pre_authorize = fields.remove("pre-authorize").unwrap();
    let tx_code_required = fields.remove("tx_code_required").unwrap();
    let send_type = fields.remove("send_type").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        CreateOfferRequest {
            credential_issuer: #credential_issuer.to_string(),
            subject_id: Some(#subject_id.to_string()),
            credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),
            pre_authorize: #pre_authorize,
            tx_code_required: #tx_code_required,
            send_type:  #send_type,
        }
    })
}

pub fn authorization_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let response_type = fields.remove("response_type").unwrap();
    let client_id = fields.remove("client_id").unwrap();
    let redirect_uri = fields.remove("redirect_uri").unwrap();
    let state = fields.remove("state").unwrap();
    //  let code_challenge = fields.remove("code_challenge").unwrap();
    let code_challenge_method = fields.remove("code_challenge_method").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        AuthorizationRequest {
            credential_issuer: #credential_issuer.to_string(),
            response_type: #response_type.to_string(),
            client_id: #client_id.to_string(),
            redirect_uri: Some(#redirect_uri.to_string()),
            state: Some(#state.to_string()),
            // code_challenge: #code_challenge.to_string(),
            code_challenge_method: #code_challenge_method.to_string(),

            // credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),

            ..Default::default()
        }
    })
}
