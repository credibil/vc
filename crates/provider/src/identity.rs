#![allow(unused)]

use std::collections::HashMap;
use std::sync::{Arc, LazyLock, Mutex};

use anyhow::{Result, anyhow, bail};
use base64ct::{Base64UrlUnpadded, Encoding};
use credibil_identity::{Key, Identity, IdentityResolver, SignerExt};
use credibil_identity::did::{
    self, Document, DocumentBuilder, KeyPurpose, PublicKeyFormat,
    VerificationMethodBuilder, VmKeyId,
};
use credibil_infosec::{Algorithm, Curve, KeyType, PublicKeyJwk};
use credibil_vc::core::generate;
use credibil_vc::format::w3c::verify;

use crate::blockstore::Mockstore;
use crate::keystore::{self, KeyUse, Keyring, SigningKey};

static DID_STORE: LazyLock<Arc<Mutex<HashMap<String, Document>>>> =
    LazyLock::new(|| Arc::new(Mutex::new(HashMap::new())));

#[derive(Clone)]
pub struct DidIdentity {
    pub url: String,
    keyring: Keyring,
}

impl DidIdentity {
    // Generate a DID-based Identity.
    pub fn new() -> Self {
        // create a new keyring and add a signing key.
        let mut keyring = Keyring::new();
        let signing_key = SigningKey::new();

        let verifying_key = PublicKeyJwk::from_bytes(&signing_key.verifying_key())
            .expect("should convert verifying key to JWK");
        keyring.add("signer", signing_key);

        // generate a did:web document
        let url = format!("https://credibil.io/{}", generate::uri_token());
        let did = did::web::default_did(&url).expect("should construct DID");

        let vk = PublicKeyJwk::from_bytes(&signing_key.verifying_key())
            .expect("should convert verifying key to JWK");
        
        let mut keyring = Keyring::new();
        keyring.add("signer", signing_key);

        let document = DocumentBuilder::new(&did)
            .add_verifying_key(&verifying_key, true)
            .expect("should add verifying key")
            .build();
        DID_STORE.lock().expect("should lock").insert(url.clone(), document);

        Self {
            url,
            keyring: keyring.clone(),
        }
    }

    pub async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        let KeyUse::Signing(signer) = self.keyring.get("signer") else {
            return Err(anyhow!("signer not found"));
        };
        Ok(signer.sign(msg))
    }

    pub async fn verifying_key(&self) -> Result<Vec<u8>> {
        let KeyUse::Signing(signer) = self.keyring.get("signer") else {
            return Err(anyhow!("signer not found"));
        };
        Ok(signer.verifying_key())
    }

    pub fn algorithm(&self) -> Algorithm {
        Algorithm::EdDSA
    }

    pub async fn verification_method(&self) -> Result<Key> {
        let Identity::DidDocument(doc) = self.resolve(&self.url).await?;
        let vm = &doc.verification_method.as_ref().unwrap()[0];
        Ok(Key::KeyId(vm.id.clone()))
    }

    pub async fn resolve(&self, url: &str) -> Result<Identity> {
        let key = url.trim_end_matches("/did.json");
        let store = DID_STORE.lock().expect("should lock");
        let Some(doc) = store.get(key).cloned() else {
            bail!("document not found");
        };
        Ok(Identity::DidDocument(doc))
    }
}
