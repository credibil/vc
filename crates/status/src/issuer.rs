//! # Status (Issuer)
//!
//! Traits and types for managing the publication of a credential status as an
//! issuer.

use std::future::Future;

pub use vercre_core::Quota;
pub use vercre_w3c_vc::model::CredentialStatus;

use crate::provider;

/// The `Status` trait specifies how status information can be looked up for a
/// credential.
pub trait Status: Send + Sync {
    /// Returns information on how to look up the status of a credential.
    ///
    /// A default implementation is provided that just returns `None` for cases
    /// where the issuer will not be providing a status endpoint.
    fn status(
        &self, _holder_subject: &str, _credential_identifier: &str,
    ) -> impl Future<Output = provider::Result<Option<Quota<CredentialStatus>>>> + Send {
        async { Ok(None) }
    }
}
