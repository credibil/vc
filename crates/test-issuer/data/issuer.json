{
    "credential_issuer": "http://credibil.io",
    "credential_endpoint": "http://credibil.io/credential",
    "deferred_credential_endpoint": "http://credibil.io/deferred",
    "display": {
        "name": "Credibil",
        "locale": "en-NZ"
    },
    "credential_configurations_supported": {
        "EmployeeID_JWT": {
            "format": "jwt_vc_json",
            "scope": "EmployeeIDCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Credibil employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/employee.png",
                        "alt_text": "Employee ID Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "http://credibil.io/assets/employee-background.png",
                        "alt_text": "Employee ID Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "EmployeeIDCredential"
                ]
            },
            "claims": [
                {
                    "path": [
                        "credentialSubject",
                        "email"
                    ],
                    "display": [
                        {
                            "name": "Email",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "family_name"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "given_name"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address"
                    ],
                    "display": [
                        {
                            "name": "Residence",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "street_address"
                    ],
                    "display": [
                        {
                            "name": "Street Address",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "locality"
                    ],
                    "display": [
                        {
                            "name": "Locality",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "region"
                    ],
                    "display": [
                        {
                            "name": "Region",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "country"
                    ],
                    "display": [
                        {
                            "name": "Country",
                            "locale": "en-NZ"
                        }
                    ]
                }
            ]
        },
        "Developer_JWT": {
            "format": "jwt_vc_json",
            "scope": "DeveloperCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Developer",
                    "description": "Credibil certified developer credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/developer.png",
                        "alt_text": "Developer Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#010100",
                    "background_image": {
                        "uri": "http://credibil.io/assets/developer-background.png",
                        "alt_text": "Developer Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "DeveloperCredential"
                ]
            },
            "claims": [
                {
                    "path": [
                        "credentialSubject",
                        "proficiency"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Proficiency",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "family_name"
                    ],
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "given_name"
                    ],
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                }
            ]
        },
        "org.iso.18013.5.1.mDL": {
            "format": "mso_mdoc",
            "doctype": "org.iso.18013.5.1.mDL",
            "scope": "DriverLicence",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Credibil employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/employee.png",
                        "alt_text": "Employee ID Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "http://credibil.io/assets/employee-background.png",
                        "alt_text": "Employee ID Background"
                    }
                }
            ],
            "claims": [
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "given_name"
                    ],
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "family_name"
                    ],
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "birth_date"
                    ],
                    "mandatory": true
                },
                {
                    "path": [
                        "org.iso.18013.5.1.aamva",
                        "organ_donor"
                    ]
                }
            ]
        }
    }
}