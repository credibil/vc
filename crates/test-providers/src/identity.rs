use std::collections::HashMap;
use std::sync::{Arc, LazyLock, Mutex};

use anyhow::{Result, bail};
use credibil_identity::did::{self, Document, DocumentBuilder};
use credibil_identity::{Identity, Key};
use credibil_se::Algorithm;

use crate::keystore::{KeyUse, Keyring};

static DID_STORE: LazyLock<Arc<Mutex<HashMap<String, Document>>>> =
    LazyLock::new(|| Arc::new(Mutex::new(HashMap::new())));

#[derive(Clone)]
pub struct DidIdentity {
    pub url: String,
    keyring: Keyring,
}

impl DidIdentity {
    // Generate a DID-based Identity.
    pub async fn new(owner: &str) -> Self {
        // create a new keyring and add a signing key.
        let mut keyring = Keyring::new(owner).await.expect("keyring created");
        keyring.add("signer", KeyUse::Signing).await.expect("signing key added");
        let verifying_key =
            keyring.verifying_key_jwk("signer").await.expect("JWK verifying key derived");

        // generate a did:web document
        let url = format!("https://credibil.io/{}", uuid::Uuid::new_v4());
        let did = did::web::default_did(&url).expect("should construct DID");

        let document = DocumentBuilder::new(&did)
            .add_verifying_key(&verifying_key, true)
            .expect("should add verifying key")
            .build();
        DID_STORE.lock().expect("should lock").insert(url.clone(), document);

        Self {
            url,
            keyring: keyring.clone(),
        }
    }

    pub async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.keyring.sign("signer", msg).await
    }

    pub async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.keyring.verifying_key("signer").await
    }

    pub fn algorithm(&self) -> Algorithm {
        Algorithm::EdDSA
    }

    pub async fn verification_method(&self) -> Result<Key> {
        let Identity::DidDocument(doc) = self.resolve(&self.url).await?;
        let vm = &doc.verification_method.as_ref().unwrap()[0];
        Ok(Key::KeyId(vm.id.clone()))
    }

    pub async fn resolve(&self, url: &str) -> Result<Identity> {
        let key = url.trim_end_matches("/did.json");
        let store = DID_STORE.lock().expect("should lock");
        let Some(doc) = store.get(key).cloned() else {
            bail!("document not found");
        };
        Ok(Identity::DidDocument(doc))
    }
}
