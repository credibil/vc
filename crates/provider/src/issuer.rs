#![allow(unused)]

use anyhow::Result;
use credibil_did::{DidResolver, Document};
use credibil_infosec::{Algorithm, Signer};
use credibil_vc::BlockStore;
use credibil_vc::status::issuer::Status;

use crate::blockstore::Mockstore;
use crate::keystore::Keyring;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone, Debug)]
pub struct Issuer {
    keyring: Keyring,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub fn new() -> Self {
        Self {
            keyring: Keyring::new(),
            blockstore: Mockstore::new(),
        }
    }
}

impl DidResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Document> {
        self.keyring.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.keyring.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.keyring.verifying_key().await
    }

    fn algorithm(&self) -> Algorithm {
        self.keyring.algorithm()
    }

    async fn verification_method(&self) -> Result<String> {
        self.keyring.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

impl Status for Issuer {}
