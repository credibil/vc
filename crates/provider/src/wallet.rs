#![allow(unused)]

use std::str::FromStr;

use anyhow::Result;
use base64ct::{Base64UrlUnpadded, Encoding};
use credibil_identity::{Identity, IdentityResolver, Key, SignerExt};
use credibil_jose::Jws;
use credibil_se::{Algorithm, Signer};
use credibil_vc::oid4vci::issuer::Credential;
use credibil_vc::oid4vp::verifier::{Claim, Queryable};
use credibil_vc::serde_cbor;
use credibil_vc::vdc::FormatProfile;
use credibil_vc::vdc::mso_mdoc::{IssuerSigned, MobileSecurityObject};
use credibil_vc::vdc::sd_jwt::SdJwtClaims;
use serde_json::Value;

use crate::blockstore::Mockstore;
use crate::identity::DidIdentity;

#[derive(Clone)]
pub struct Wallet {
    identity: DidIdentity,
    // blockstore: Mockstore,
    store: Vec<Queryable>,
}

impl Wallet {
    pub fn new() -> Self {
        Self {
            identity: DidIdentity::new(),
            // blockstore: Mockstore::new(),
            store: Vec::new(),
        }
    }

    // Add a credential to the store.
    pub fn add(&mut self, queryable: Queryable) {
        self.store.push(queryable);
    }

    pub fn fetch(&self) -> &[Queryable] {
        &self.store
    }
}

impl IdentityResolver for Wallet {
    async fn resolve(&self, url: &str) -> anyhow::Result<Identity> {
        self.identity.resolve(url).await
    }
}

impl Signer for Wallet {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.identity.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.identity.verifying_key().await
    }

    async fn algorithm(&self) -> Result<Algorithm> {
        Ok(self.identity.algorithm())
    }
}

impl SignerExt for Wallet {
    async fn verification_method(&self) -> Result<Key> {
        self.identity.verification_method().await
    }
}
