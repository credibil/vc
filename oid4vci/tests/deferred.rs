//! Deferred Issuance Tests

use std::collections::HashMap;

use credibil_oid4vci::blockstore::BlockStore;
use credibil_oid4vci::identity::{Key, SignerExt};
use credibil_oid4vci::jose::{JwsBuilder, Jwt, decode_jws};
use credibil_oid4vci::proof::W3cVcClaims;
use credibil_oid4vci::types::{
    CreateOfferRequest, Credential, CredentialRequest, CredentialResponse, Dataset,
    DeferredCredentialRequest, NonceRequest, ProofClaims, TokenGrantType, TokenRequest,
};
use credibil_oid4vci::{CredentialHeaders, DeferredHeaders, JwtType, OneMany, did_jwk};
use serde_json::json;
use test_utils::issuer::{CAROL_ID, ISSUER_ID, Issuer, data};
use test_utils::wallet::Wallet;
use tokio::sync::OnceCell;

static CAROL: OnceCell<Wallet> = OnceCell::const_new();
async fn carol() -> &'static Wallet {
    CAROL.get_or_init(|| async { Wallet::new("https://deferred.io/carol").await }).await
}

// Should return a credential when using the pre-authorized code flow and the
// credential offer to the Wallet is made by value.
#[tokio::test]
async fn deferred() {
    let provider = Issuer::new("https://deferred.io/deferred").await;
    let carol = carol().await;

    BlockStore::put(&provider, "owner", "ISSUER", ISSUER_ID, data::ISSUER).await.unwrap();
    BlockStore::put(&provider, "owner", "SERVER", ISSUER_ID, data::SERVER).await.unwrap();
    BlockStore::put(&provider, "owner", "SUBJECT", CAROL_ID, data::PENDING_USER).await.unwrap();

    // --------------------------------------------------
    // Alice creates a credential offer for Bob
    // --------------------------------------------------
    let request = CreateOfferRequest::builder()
        .subject_id(CAROL_ID)
        .with_credential("EmployeeID_W3C_VC")
        .build();
    let response =
        credibil_oid4vci::handle(ISSUER_ID, request, &provider).await.expect("should create offer");

    // --------------------------------------------------
    // Bob receives the offer and requests a token
    // --------------------------------------------------
    let offer = response.offer_type.as_object().expect("should have offer").clone();
    let grants = offer.grants.expect("should have grant");
    let pre_auth_grant = grants.pre_authorized_code.expect("should have pre-authorized code grant");

    let request = TokenRequest::builder()
        .grant_type(TokenGrantType::PreAuthorizedCode {
            pre_authorized_code: pre_auth_grant.pre_authorized_code,
            tx_code: response.tx_code.clone(),
        })
        .build();
    let token =
        credibil_oid4vci::handle(ISSUER_ID, request, &provider).await.expect("should return token");

    // --------------------------------------------------
    // Bob receives the token and prepares a proof for a credential request
    // --------------------------------------------------
    let nonce = credibil_oid4vci::handle(ISSUER_ID, NonceRequest, &provider)
        .await
        .expect("should return nonce");

    // proof of possession of key material
    let key = carol
        .verification_method()
        .await
        .expect("should have did")
        .try_into()
        .expect("should map key to key binding");

    let jws = JwsBuilder::new()
        .typ(JwtType::ProofJwt)
        .payload(ProofClaims::new().credential_issuer(ISSUER_ID).nonce(&nonce.c_nonce))
        .key_ref(&key)
        .add_signer(carol)
        .build()
        .await
        .expect("builds JWS");
    let jwt = jws.encode().expect("encodes JWS");

    // --------------------------------------------------
    // Bob requests a credential and receives a deferred response
    // --------------------------------------------------
    let details = token.authorization_details.as_ref().expect("should have authorization details");
    let request = CredentialRequest::builder()
        .credential_identifier(&details[0].credential_identifiers[0])
        .with_proof(jwt)
        .build();

    let request = credibil_oid4vci::Request {
        body: request,
        headers: CredentialHeaders {
            authorization: token.access_token.clone(),
        },
    };

    let response = credibil_oid4vci::handle(ISSUER_ID, request, &provider)
        .await
        .expect("should return credential");

    // --------------------------------------------------
    // Alice approves issuance of the credential
    // HACK: update subject's pending state
    // --------------------------------------------------
    let credential_identifier = &details[0].credential_identifiers[0];
    let mut subject: HashMap<String, Dataset> = serde_json::from_slice(data::PENDING_USER).unwrap();
    let mut credential: Dataset = subject.get(credential_identifier).unwrap().clone();
    credential.pending = false;

    subject.insert(credential_identifier.to_string(), credential);
    let data = serde_json::to_vec(&subject).unwrap();

    BlockStore::delete(&provider, "owner", "SUBJECT", CAROL_ID).await.unwrap();
    BlockStore::put(&provider, "owner", "SUBJECT", CAROL_ID, &data).await.unwrap();

    // --------------------------------------------------
    // After a brief wait Bob retrieves the credential
    // --------------------------------------------------
    let CredentialResponse::TransactionId { transaction_id } = &*response else {
        panic!("expected transaction_id");
    };

    let request = credibil_oid4vci::Request {
        body: DeferredCredentialRequest {
            transaction_id: transaction_id.clone(),
        },
        headers: DeferredHeaders {
            authorization: token.access_token.clone(),
        },
    };
    let response = credibil_oid4vci::handle(ISSUER_ID, request, &provider)
        .await
        .expect("should return credential");

    // --------------------------------------------------
    // Bob extracts and verifies the received credential
    // --------------------------------------------------
    let CredentialResponse::Credentials { credentials, .. } = &*response else {
        panic!("expected single credential");
    };

    let Credential { credential } = credentials.first().expect("should have credential");

    // verify the credential proof
    let token = credential.as_str().expect("should be a string");
    let resolver = async |kid: String| did_jwk(&kid, &provider).await;
    let jwt: Jwt<W3cVcClaims> = decode_jws(token, resolver).await.expect("should decode");

    // verify the credential
    let Key::KeyId(carol_kid) = carol.verification_method().await.unwrap() else {
        panic!("should have did");
    };
    let carol_did = carol_kid.split('#').next().expect("should have did");

    assert_eq!(jwt.claims.iss, ISSUER_ID);
    assert_eq!(jwt.claims.sub, carol_did);

    let OneMany::One(subject) = jwt.claims.vc.credential_subject else {
        panic!("should be a single credential subject");
    };
    assert_eq!(subject.id, Some(carol_did.to_string()));
    assert_eq!(subject.claims.get("family_name"), Some(&json!("Person")));
}
