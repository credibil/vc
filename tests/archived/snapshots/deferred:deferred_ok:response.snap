---
source: tests/deferred.rs
assertion_line: 96
expression: "&response"
---
credential: "[credential]"
c_nonce: "[c_nonce]"
c_nonce_expires_in: 599
notification_id: "[notification_id]"
